Python/Product/PythonTools/PythonTools/Navigation/Navigable/NavigableSymbolSource.cs

```diff
@@ -0,0 +1,103 @@
+namespace Microsoft.PythonTools.Navigation.Navigable {
+    class NavigableSymbolSource : INavigableSymbolSource {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ITextView _textView;
+        private readonly ITextBuffer _buffer;
+        private readonly IClassifier _classifier;
+        private readonly ITextStructureNavigator _textNavigator;
+        private readonly AnalysisEntryService _entryService;
+
+        private static readonly string[] _classifications = new string[] {
+            PredefinedClassificationTypeNames.Identifier,
+            PythonPredefinedClassificationTypeNames.Class,
+            PythonPredefinedClassificationTypeNames.Function,
+            PythonPredefinedClassificationTypeNames.Module,
+            PythonPredefinedClassificationTypeNames.Parameter,
+        };
+
+        public NavigableSymbolSource(IServiceProvider serviceProvider, ITextView textView, ITextBuffer buffer, IClassifier classifier, ITextStructureNavigator textNavigator) {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            _textView = textView ?? throw new ArgumentNullException(nameof(textView));
+            _buffer = buffer ?? throw new ArgumentNullException(nameof(buffer));
+            _classifier = classifier ?? throw new ArgumentNullException(nameof(classifier));
+            _textNavigator = textNavigator ?? throw new ArgumentNullException(nameof(textNavigator));
+            _entryService = serviceProvider.GetEntryService();
+        }
+
+        public void Dispose() {
+        }
+
+        public async Task<INavigableSymbol> GetNavigableSymbolAsync(SnapshotSpan triggerSpan, CancellationToken cancellationToken) {
+            Debug.Assert(triggerSpan.Length == 1);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var extent = _textNavigator.GetExtentOfWord(triggerSpan.Start);
+            if (!extent.IsSignificant) {
+                return null;
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            AnalysisEntry entry = null;
+            _entryService?.TryGetAnalysisEntry(_textView, _buffer, out entry);
+            if (entry == null) {
+                return null;
+            }
+
+            foreach (var token in _classifier.GetClassificationSpans(extent.Span)) {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Quickly eliminate anything that isn't the right classification.
+                var name = token.ClassificationType.Classification;
+                if (!_classifications.Any(c => name.Contains(c))) {
+                    continue;
+                }
+
+                // Check with the analyzer, which will give us a precise
+                // result, including the source location.
+                var result = await entry.Analyzer.AnalyzeExpressionAsync(entry, _textView, token.Span.Start);
```

Comment:
Cancellation is likely for this function, so we should throw here if the token has been cancelled (since this is the longest-running operation).

I suspect if it's cancelled then the result will be ignored anyway, but no point doing extra work. We should also use `ConfigureAwait` on the AnalyzeExpressionAsync call to avoid marshalling to a particular thread on return - the caller will do that if necessary.